src/software/ai/hl/stp/play/free_kick/free_kick_play_fsm.cpp

```diff
@@ -0,0 +1,245 @@
+    for (unsigned int i = 0; i < offensive_positioning_tactics.size(); i++)
+    {
+        Zones zone = {ranked_zones[i]};
+        auto pass  = pass_eval.getBestPassInZones(zone).pass;
+
+        offensive_positioning_tactics[i]->updateControlParams(
+            pass.receiverPoint(), pass.receiverOrientation(), 0.0);
+    }
+}
+
+void FreeKickPlayFSM::updateAlignToBallTactic(const WorldPtr &world_ptr)
+{
+    // We want the kicker to get into position behind the ball facing the enemy net
+    Point ball_pos                = world_ptr->ball().position();
+    Vector ball_to_enemy_goal_vec = world_ptr->field().enemyGoalCenter() - ball_pos;
+    align_to_ball_tactic->updateControlParams(
+        ball_pos - ball_to_enemy_goal_vec.normalize(ROBOT_MAX_RADIUS_METERS * 2),
+        ball_to_enemy_goal_vec.orientation(), 0);
+}
+
+bool FreeKickPlayFSM::shotFound(const Update &event)
+{
+    shot = calcBestShotOnGoal(event.common.world_ptr->field(),
+                              event.common.world_ptr->friendlyTeam(),
+                              event.common.world_ptr->enemyTeam(),
+                              event.common.world_ptr->ball().position(), TeamType::ENEMY);
+    return shot &&
+           shot->getOpenAngle() >
+               Angle::fromDegrees(
+                   ai_config.attacker_tactic_config().min_open_angle_for_shot_deg());
+}
+
+void FreeKickPlayFSM::shootBall(const Update &event)
+{
+    LOG(DEBUG) << "Shooting ball...";
+    PriorityTacticVector tactics_to_run = {{}};
+
+    Point ball_pos = event.common.world_ptr->ball().position();
+
+    shoot_tactic->updateControlParams(
+        ball_pos, (shot->getPointToShootAt() - ball_pos).orientation(),
+        BALL_MAX_SPEED_METERS_PER_SECOND);
+    tactics_to_run[0].emplace_back(shoot_tactic);
+
+    event.common.set_tactics(tactics_to_run);
+}
+
+void FreeKickPlayFSM::startLookingForPass(const FreeKickPlayFSM::Update &event)
+{
+    pass_optimization_start_time = event.common.world_ptr->getMostRecentTimestamp();
+    // Generate the best zones for receiving a pass
+    // Only generate the zones once to avoid oscillatory behaviour
+    ranked_zones =
+        pass_generator.generatePassEvaluation(*event.common.world_ptr)
+            .rankZonesForReceiving(*event.common.world_ptr,
+                                   best_pass_and_score_so_far.pass.receiverPoint());
+}
+
+bool FreeKickPlayFSM::timeExpired(const FreeKickPlayFSM::Update &event)
+{
+    Duration time_since_pass_optimization_start =
+        event.common.world_ptr->getMostRecentTimestamp() - pass_optimization_start_time;
+    return time_since_pass_optimization_start.toSeconds() >
+           ai_config.free_kick_play_config().max_time_commit_to_pass_seconds();
+}
+
+void FreeKickPlayFSM::chipBall(const Update &event)
+{
+    LOG(DEBUG) << "Time to look for pass expired. Chipping ball...";
+    PriorityTacticVector tactics_to_run = {{}};
+
+    Point chip_target = event.common.world_ptr->field().enemyGoalCenter();
```

Comment:
Could we chip towards the friendly which is the most up the field instead (robot with highest x-value that is not the robot with the ball doing the freekick)?